### DISCUSSION TITLE
How to give the focused ListBox item a focus rectangle

### DISCUSSION THREAD

**Discussion by SimonORorke** [AvaloniaUI/Avalonia#16173] [Q&A] [answered]
How can I give the focused `ListBox` item a focus rectangle? I tried this, but it does nothing:
```xml
<Style Selector="ListBoxItem">
    <Setter Property="FocusAdorner">
        <FocusAdornerTemplate>
            <Border BorderThickness="2"
                    BorderBrush="White">
            </Border>
        </FocusAdornerTemplate>
    </Setter>
</Style>
```
#10158 and #10415 appear to be relevant. But I don't know how to proceed.

**Comment by stevemonaco** [comment]
This works for me. AFAIK, focus adorners are only shown on keyboard navigation. Given this demo:

```xml
<Window.Styles>
    <Style Selector="ListBoxItem">
        <Setter Property="FocusAdorner">
            <FocusAdornerTemplate>
                <Border BorderBrush="Orange" BorderThickness="2" />
            </FocusAdornerTemplate>
        </Setter>
    </Style>
</Window.Styles>

<Grid RowDefinitions="*,auto">
    <ListBox>
        <ListBoxItem>Item 1</ListBoxItem>
        <ListBoxItem>Item 2</ListBoxItem>
        <ListBoxItem>Item 3</ListBoxItem>
        <ListBoxItem>Item 4</ListBoxItem>
    </ListBox>
    <TextBox Grid.Row="1" Text="Testing" />
</Grid>
```

![listboxadorner](https://github.com/AvaloniaUI/Avalonia/assets/1782158/1a060298-fa76-4ed3-87e2-bd2b8491efd7)

You can see the adorner only shows when using tab navigation (which sets `:focus-visible` only in `OnGotFocus`). When I change the selected item via up/down arrow keys, it goes away.

Since maintaining a focus adorner with arrow key navigation is going against intended design, I assume, there's two approaches: 1. set `FocusAdorner` to `{x:Null}` and show a border when `:selected`. 2. write something hacky like below to unfocus/refocus:

```cs
private void Lb_SelectionChanged(object? sender, SelectionChangedEventArgs e)
{
    // Add better error handling in real code
    FocusManager.ClearFocus();
    var item = (ListBoxItem)e.AddedItems[0];
    item.Focus(Avalonia.Input.NavigationMethod.Tab);
}
```

**Reply by SimonORorke** [reply]
Thanks @stevemonaco!

I find it impossible to navigate the list with the Tab key.  I wonder why you can and I can't? However, that is moot, as your `SelectionChanged` has given me the right idea.  For me, it does not quite work as you have coded, since my `Items` are `string`s (because bound to a view model list of strings), whereas in your example  `Items` are `ListBoxItem`s.

So what has worked for me is to get rid of the `Style` and  add this variant of the `SelectionChanged` event handler you suggested:
```csharp
private void ListBoxOnSelectionChanged(object? sender, SelectionChangedEventArgs e) {
    var listBoxItems = (
        from descendant in ColourSchemeListBox.GetVisualDescendants()
        where descendant is ListBoxItem
        select (ListBoxItem)descendant).ToList();
    if (IsVisible) {
       var selectedListBoxItem = listBoxItems[ColourSchemeListBox.SelectedIndex];
       selectedListBoxItem.Focus(NavigationMethod.Directional);
    }
}
```

**Reply by stevemonaco** [reply, marked as answer]
I'm not sure why tab navigation (to the `ListBox` itself, not between the items within) doesn't repro on your end. I'm using Win11 / 11.1rc-1 / Fluent theme.

I strongly suggest to avoid copying descendants and instead:

```cs
var item = (ListBoxItem) ColourSchemeListBox.ContainerFromIndex(ColourSchemeListBox.SelectedIndex);
```

**Reply by SimonORorke** [reply]
Re tab navigation:  Ah, you mean to the `ListBox` itself, not between the items within. Yes, I can do that.

I did not know about `ContainerFromIndex`.  That's a good trick!  So here's my revised `OnSelectionChanged` event handle:
```csharp
private void ListBoxOnSelectionChanged(
    object? sender, SelectionChangedEventArgs e) {
    if (IsVisible) {
        var selectedListBoxItem = (ListBoxItem)ListBox.ContainerFromIndex(ListBox.SelectedIndex)!;
        // NavigationMethod.Directional puts a focus rectangle round the selected item.
        selectedListBoxItem.Focus(NavigationMethod.Directional);
    }
}
```

**Reply by stevemonaco** [reply]
You should safeguard against `SelectedIndex == -1` for when a selection is cleared. Looks ok otherwise.

Edit: This method actually has some bounds guards (returning null), so you could safeguard against null instead.

https://github.com/AvaloniaUI/Avalonia/blob/3b235f9aa82dbcae024c0e75486e7d91df1be13d/src/Avalonia.Controls/Presenters/ItemsPresenter.cs#L188-L192

**Reply by SimonORorke** [reply]
> You should safeguard against `SelectedIndex == -1`

Good point!  I left that out because, in my application, there's code I did not show you that automatically selects and focuses an item when the window has loaded, so there's never not a selected item. To make the code more generally useful for anyone else who might be interested, here's an update:
```csharp
private void ListBoxOnSelectionChanged(
    object? sender, SelectionChangedEventArgs e) {
    if (IsVisible && ListBox.SelectedIndex >= 0) {
        var selectedListBoxItem = (ListBoxItem)ListBox.ContainerFromIndex(ListBox.SelectedIndex)!;
        // NavigationMethod.Directional puts a focus rectangle round the selected item.
        selectedListBoxItem.Focus(NavigationMethod.Directional);
    }
}
```
